File: modules/agent/src/serializers/provenance.py
"""Serializers for provenance-related models."""


def provenance_to_response(p) -> dict:
    """Convert provenance model to response dict."""
    return {
        "id": p.id,
        "entity_type": p.entity_type,
        "entity_id": p.entity_id,
        "field_name": p.field_name,
        "source": p.source,
        "source_url": p.source_url,
        "confidence": float(p.confidence) if p.confidence else None,
        "verified_at": p.verified_at.isoformat() if p.verified_at else None,
        "verified_by": p.verified_by,
        "raw_value": p.raw_value,
        "created_at": p.created_at.isoformat() if p.created_at else None,
        "updated_at": p.updated_at.isoformat() if p.updated_at else None,
    }

File: modules/agent/internal/serializers/provenance.go
package serializers

import (
	"time"

	"github.com/pina-colada-co/agent-go/internal/models"
	"github.com/shopspring/decimal"
	"gorm.io/datatypes"
)

// ProvenanceResponse represents a data provenance record
type ProvenanceResponse struct {
	ID         int64            `json:"id"`
	EntityType string           `json:"entity_type"`
	EntityID   int64            `json:"entity_id"`
	FieldName  string           `json:"field_name"`
	Source     string           `json:"source"`
	SourceURL  *string          `json:"source_url"`
	Confidence *decimal.Decimal `json:"confidence"`
	VerifiedAt time.Time        `json:"verified_at"`
	VerifiedBy *int64           `json:"verified_by"`
	RawValue   datatypes.JSON   `json:"raw_value"`
	CreatedAt  time.Time        `json:"created_at"`
	UpdatedAt  time.Time        `json:"updated_at"`
}

// ProvenanceToResponse converts a model to response
func ProvenanceToResponse(m *models.DataProvenance) ProvenanceResponse {
	return ProvenanceResponse{
		ID:         m.ID,
		EntityType: m.EntityType,
		EntityID:   m.EntityID,
		FieldName:  m.FieldName,
		Source:     m.Source,
		SourceURL:  m.SourceURL,
		Confidence: m.Confidence,
		VerifiedAt: m.VerifiedAt,
		VerifiedBy: m.VerifiedBy,
		RawValue:   m.RawValue,
		CreatedAt:  m.CreatedAt,
		UpdatedAt:  m.UpdatedAt,
	}
}

// ProvenancesToResponse converts a slice of models to responses
func ProvenancesToResponse(models []models.DataProvenance) []ProvenanceResponse {
	resp := make([]ProvenanceResponse, len(models))
	for i := range models {
		resp[i] = ProvenanceToResponse(&models[i])
	}
	return resp
}
